inst/include/Rcpp/sugar/functions/median.h

```diff
@@ -0,0 +1,296 @@
+// -*- mode: C++; c-indent-level: 4; c-basic-offset: 4; indent-tabs-mode: nil; -*-
+//
+// median.h: Rcpp R/C++ interface class library -- median
+//
+// Copyright (C) 2016 Dirk Eddelbuettel, Romain Francois, and Nathan Russell
+//
+// This file is part of Rcpp.
+//
+// Rcpp is free software: you can redistribute it and/or modify it
+// under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 2 of the License, or
+// (at your option) any later version.
+//
+// Rcpp is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Rcpp.  If not, see <http://www.gnu.org/licenses/>.
+
+#ifndef Rcpp__sugar__median_h
+#define Rcpp__sugar__median_h
+
+namespace Rcpp {
+namespace sugar {
+namespace median_detail {
+
+// need to return double for integer vectors 
+// (in case of even-length input vector)
+// and Rcpp::String for STRSXP
+// also need to return NA_REAL for 
+// integer vector yielding NA result
+template <int RTYPE>
+struct result {
+    typedef typename Rcpp::traits::storage_type<RTYPE>::type type;
+    enum { rtype = RTYPE };
+};
+
+template <>
+struct result<INTSXP> {
+    typedef double type;
+    enum { rtype = REALSXP };
+};
+
+template <>
+struct result<STRSXP> {
+    typedef Rcpp::String type;
+    enum { rtype = STRSXP };
+};
+
+// std::nth_element and std::max_element don't 
+// know how to compare Rcomplex values
+template <typename T>
+inline bool less(T lhs, T rhs) {
+    return lhs < rhs;
+}
+
+template<>
+inline bool less<Rcomplex>(Rcomplex lhs, Rcomplex rhs) {
+    if (lhs.r < rhs.r) return true;
+    if (lhs.i < rhs.i) return true;
+    return false; 
+}
+
+// compiler does not know how to handle 
+// Rcomplex numerator / double denominator
+// and need explicit cast for INTSXP case
+inline double half(double lhs) {
+    return lhs / 2.0;
+}
+
+inline double half(int lhs) {
+    return static_cast<double>(lhs) / 2.0;
+}
+
+inline Rcomplex half(Rcomplex lhs) {
+    lhs.r /= 2.0;
+    lhs.i /= 2.0;
+    return lhs;
+}
+
+} // median_detail
+
+// base case
+template <int RTYPE, bool NA, typename T, bool NA_RM = false>
+class Median {
+public:
+    typedef typename median_detail::result<RTYPE>::type result_type;
+    typedef typename Rcpp::traits::storage_type<RTYPE>::type stored_type;
+    enum { RTYPE2 = median_detail::result<RTYPE>::rtype };
```

Comment:
Maybe instead of `RTYPE2` we could use `RESULT_RTYPE` or `result_rtype` or something?